lib/environment.cpp

```diff
@@ -0,0 +1,86 @@
+
+#include "shell/environment.hpp"
+
+#include <cstring>
+
+namespace shell {
+
+EnvironmentVariables::EnvironmentVariables(const map<string, string> &variables) : envp(nullptr) {
+	const auto listSize = variables.size() + 1;
+	try {
+		envp = new char *[listSize];
+
+		size_t i = 0;
+		for (const auto &[name, value] : variables) {
+			envp[i] = nullptr;
+			char *variable = new char[name.size() + value.length() + 2]; // +2 for '=' and '/0'
+			std::memmove(variable, name.data(), name.length());
+			std::memmove(variable + name.length(), "=", 1);
+			std::memmove(variable + name.length() + 1, value.data(), value.length() + 1); // +1 to include '/0'
+			i++;
+		}
+	} catch (const std::exception &e) {
+		destroy();
+		throw e;
+	}
+}
+
+EnvironmentVariables::~EnvironmentVariables() {
+	destroy();
+}
+
+void EnvironmentVariables::destroy() {
+	if (envp) {
+		while (*envp != nullptr) {
+			delete[] (*envp);
+			envp++;
+		}
+		delete[] envp;
+		envp = nullptr;
+	}
+}
+
+char *const *EnvironmentVariables::raw() const {
+	return envp;
+}
+
+Environment &Environment::getInstance() {
+	static Environment environment;
+	return environment;
+}
+
+void Environment::setEnvironmentVariables(char *const *envp) {
+	auto &environment = getInstance();
+	for (; *envp != nullptr; envp++) {
+		const string variable = *envp;
+
+		const auto splitPoint = variable.find('=');
+		assert(splitPoint != string::npos);
+
+		const auto name = variable.substr(0, splitPoint);
+		const auto value = variable.substr(splitPoint + 1);
+		environment.variables[name] = value;
+	}
+}
+
+EnvironmentVariables Environment::getEnvironmentVariables() {
+	const auto &environment = getInstance();
```

Comment:
Are we allowed to use the library environment functions like `getenv` etc? Would it make sense to make the Environment class a wrapper around that interface instead? (if the interface is nice, maybe we could use it directly?)

Reply:
Actually not a bad idea ✔️

Reply:
We might still need some custom environment implementation, as I think there's a different between exported and unexported variables (i.e. `a=b` vs `export a=b`). I guess this should be something like a symbol table as there will be things like functions as well

Reply:
I'm not sure about the symbol table, I don't think functions become a part of this, they are a separate concept
But yes `a=b` becomes an environment variable only for the currently executed command.

Basically a transient environment variable, not set in the shell-global environment

Reply:
Apparently exported functions are also simply stored as strings: https://unix.stackexchange.com/questions/22796/can-i-export-functions-in-bash
The 'transient' environment could also technically store these functions I suppose, I don't think there are many other symbols so might not be worth it to create a variant or something like that.